File: modules/javafx.graphics/src/main/native-prism-d3d11/D3D11Context.cc
/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include <iostream>
#include <comdef.h>

#include "D3D11Pipeline.h"
#include "D3D11Context.h"
#include "D3D11PipelineManager.h"

#include "com_sun_prism_d3d11_D3D11Context.h"

// static
HRESULT
D3D11Context::CreateInstance(UINT adapter, bool isVsyncEnabled, D3D11Context **ppCtx) {
    fprintf(stderr, "D3D11Context::CreateInstance\n");
    HRESULT res;
    *ppCtx = new D3D11Context(adapter);
    if (FAILED(res = (*ppCtx)->InitContext(isVsyncEnabled))) {
        fprintf(stderr, "InitContext failed!");
        delete *ppCtx;
        *ppCtx = NULL;
    }
    fprintf(stderr, "returning res\n");
    return res;
}

D3D11Context::D3D11Context(UINT adapter) {
    pd3d11Device = NULL;
    pd3d11DevCtx = NULL;
    adapterOrdinal = adapter;
    pResourceMgr = NULL;
    ZeroMemory(&featureLevel, sizeof(D3D_FEATURE_LEVEL));
}

HRESULT D3D11Context::InitContext(bool isVsyncEnabled) {
    fprintf(stderr, "D3D11Context::InitContext\n");
    D3D_FEATURE_LEVEL lvl[] = {
        D3D_FEATURE_LEVEL_11_1, D3D_FEATURE_LEVEL_11_0,
        D3D_FEATURE_LEVEL_10_1, D3D_FEATURE_LEVEL_10_0,
        D3D_FEATURE_LEVEL_9_3, D3D_FEATURE_LEVEL_9_2, D3D_FEATURE_LEVEL_9_1 };

    DWORD createDeviceFlags = D3D11_CREATE_DEVICE_BGRA_SUPPORT;
    #ifdef D3D_DEBUG_INFO
        fprintf(stderr, "D3D_DEBUG_INFO set!\n");
        createDeviceFlags |= D3D11_CREATE_DEVICE_DEBUG;
    #endif

    D3D_FEATURE_LEVEL fl;
    // Note: If prismd3d11.dll crashes here, and DEBUG is set, make sure to enabled
    // the DirectX developer runtime. On Windows 10 that means enabling optional Windows
    // feature "Graphics Tools" (search for "Manage Optional Features").
    HRESULT hr = D3D11CreateDevice(NULL, D3D_DRIVER_TYPE_HARDWARE, NULL,
        createDeviceFlags, lvl, _countof(lvl),
        D3D11_SDK_VERSION, &pd3d11Device, &fl, &pd3d11DevCtx);
    if (hr == E_INVALIDARG) {
        // DirectX 11.1 is not supported - fall back to 11.0.
        fprintf(stderr, "Falling back to 11.0\n");
        hr = D3D11CreateDevice(NULL, D3D_DRIVER_TYPE_HARDWARE, NULL,
            createDeviceFlags, &lvl[1], _countof(lvl) - 1,
            D3D11_SDK_VERSION, &pd3d11Device, &fl, &pd3d11DevCtx);
    }

    if (FAILED(hr) || pd3d11Device == NULL || pd3d11DevCtx == NULL) {
        _com_error error(hr);
        LPCTSTR errorText = error.ErrorMessage();
        std::wstring w_err((wchar_t*)errorText);
        std::wcerr << "D3D11CreateDevice FAILED: '" << errorText << "'" << std::endl;
        // SetErrorMessage("InitD3D11: unable to create IDirect3D11 device");
        fprintf(stderr, "InitD3D11: unable to create IDirect3D11 device\n");
        return E_FAIL;
    } else {
        fprintf(stderr, "IDirect3D11 device created successfully!\n");
    }

    if (pResourceMgr == NULL) {
        pResourceMgr = D3D11ResourceManager::CreateInstance(this);
    }
    return hr;
}

HRESULT
D3D11Context::ResetClip() {
    fprintf(stderr, "D3D11Context::ResetClip\n");
    if (pd3d11Device == NULL) {
        fprintf(stderr, "pd3d11Device was NULL\n");
        return E_FAIL;
    }

    D3D11_RASTERIZER_DESC rsDesc;
    ZeroMemory(&rsDesc, sizeof(D3D11_RASTERIZER_DESC));
    rsDesc.ScissorEnable = TRUE;
    rsDesc.FillMode = D3D11_FILL_SOLID;
    rsDesc.CullMode = D3D11_CULL_BACK;
    ID3D11RasterizerState* rsState;
    HRESULT hr = pd3d11Device->CreateRasterizerState(&rsDesc, &rsState);
    if (FAILED(hr)) {
        _com_error error(hr);
        LPCTSTR errorText = error.ErrorMessage();
        std::wstring w_err((wchar_t*)errorText);
        std::wcerr << "CreateRasterizerState FAILED: '" << errorText << "'" << std::endl;
        return hr;
    }
    pd3d11DevCtx->RSSetState(rsState);
    rsState->Release();
    return hr;
}

HRESULT
D3D11Context::ResetTransform() {
    if (pd3d11Device == NULL) {
        return E_FAIL;
    }

    DirectX::XMStoreFloat4x4(&world, DirectX::XMMatrixIdentity());
    return UpdateVertexShaderTX();
}


HRESULT
D3D11Context::UpdateVertexShaderTX() {
    if (pd3d11Device == NULL) {
        return E_FAIL;
    }

    DirectX::XMFLOAT4X4 wvp;
    // create the WorldViewProj matrix
    // wvp = T(w * v * p);
    // since view is currently included in the projection matrix, wvp = T(w * p)
    DirectX::XMStoreFloat4x4(&wvp, XMMatrixMultiplyTranspose(XMLoadFloat4x4(&world), XMLoadFloat4x4(&projection)));
    // Apply the pixel adjustment values for the current render target.
    // These values adjust our default (identity) coordinates so that the
    // pixel edges are at integer coordinate locations.

    // FIXME.
    // wvp._14 += pixadjustx;
    // wvp._24 += pixadjusty;
    m_constantBufferData.mWorldViewProj = wvp;
    CD3D11_BUFFER_DESC cbDesc(sizeof(ConstantBufferStruct), D3D11_BIND_CONSTANT_BUFFER);
    HRESULT hr = pd3d11Device->CreateBuffer(&cbDesc, NULL, &pd3d11ConstantBuffer);
    if (FAILED(hr)) {
        return hr;
    }
    pd3d11DevCtx->VSSetConstantBuffers(0, 1, &pd3d11ConstantBuffer);
    return hr;
}

/**
 * This method releases context resources either from the default pool only
 * (basically from vram) or all of them, depending on the passed argument.
 *
 * Note that some resources are still not under ResourceManager control so we
 * have to handle them separately. Ideally we'd move every allocated resource
 * under RM control.
 *
 * The reason we have single method instead of a pair of methods (one for
 * default only and one for everything) is to reduce code duplication. It is
 * possible to call ReleaseDefPoolResources from ReleaseContextResources but
 * then we'd traverse the resources list twice (may not be a big deal).
 */
void D3D11Context::ReleaseContextResources(int releaseType) {
    fprintf(stderr, "D3D11Context::ReleaseContextResources\n");
    /*
    TraceLn2(NWT_TRACE_INFO,
             "D3DContext::ReleaseContextResources: %d pd3dDevice = 0x%x",
             releaseType, pd3dDevice);

    if (releaseType != RELEASE_ALL && releaseType != RELEASE_DEFAULT) {
        TraceLn1(NWT_TRACE_ERROR,
                "D3DContext::ReleaseContextResources unknown type: %d",
                releaseType);
        return;
    }

    EndScene();

    if (releaseType == RELEASE_DEFAULT) {
        if (pVertexBufferRes != NULL && pVertexBufferRes->IsDefaultPool()) {
            // if VB is in the default pool it will be released by the RM
            pVertexBufferRes = NULL;
        }
        pResourceMgr->ReleaseDefPoolResources();
    */
    if (releaseType == RELEASE_ALL) {
        // will be released with the resource manager
        // pVertexBufferRes = NULL;
        // SAFE_RELEASE(pVertexDecl);
        // SAFE_RELEASE(pIndices);
        // SAFE_RELEASE(pPassThroughVS);
        SAFE_DELETE(pResourceMgr);
    }
}

int D3D11Context::release() {
    fprintf(stderr, "D3D11Context::release\n");
    // TraceLn2(NWT_TRACE_INFO, "~D3DContext: pd3dDevice=0x%x, pd3dObject =0x%x", pd3dDevice, pd3dObject);
    ReleaseContextResources(RELEASE_ALL);
    /*
    for (int i = 0; i < NUM_TEXTURE_CACHE; i++) {
        SAFE_RELEASE(textureCache[i].surface);
        SAFE_RELEASE(textureCache[i].texture);
    }
    */
    SAFE_RELEASE(pd3d11Device);

    /*
    if (phongShader) {
        delete phongShader;
        phongShader = NULL;
    }
    */
    delete this;
    return 0;
}

File: modules/javafx.graphics/src/main/java/com/sun/prism/d3d11/D3D11Context.java
/*
 * Copyright (c) 2008, 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package com.sun.prism.d3d11;

import com.sun.glass.ui.Screen;
import com.sun.javafx.geom.Rectangle;
import com.sun.javafx.geom.transform.BaseTransform;
import com.sun.javafx.sg.prism.NGCamera;
import com.sun.prism.CompositeMode;
import com.sun.prism.RTTexture;
import com.sun.prism.RenderTarget;
import com.sun.prism.Texture;
import com.sun.prism.impl.PrismSettings;
import com.sun.prism.impl.ps.BaseShaderContext;
import com.sun.prism.ps.Shader;

public class D3D11Context extends BaseShaderContext {

    private final D3D11ResourceFactory factory;
    public static final int NUM_QUADS = PrismSettings.superShader ? 4096 : 256;
    private final long pContext;
    private State state;

    D3D11Context(long pContext, Screen screen, D3D11ResourceFactory factory) {
        super(screen, factory, NUM_QUADS);
        this.pContext = pContext;
        this.factory = factory;
    }

    @Override
    public D3D11ResourceFactory getResourceFactory() {
        return factory;
    }

    protected void initState() {
        init();
        state = new State();
    }

    long getContextHandle() {
        return pContext;
    }

    @Override
    protected State updateRenderTarget(RenderTarget target, NGCamera camera, boolean depthTest) {
        return null;
    }

    @Override
    protected void updateTexture(int texUnit, Texture tex) {

    }

    @Override
    protected void updateShaderTransform(Shader shader, BaseTransform xform) {

    }

    @Override
    protected void updateWorldTransform(BaseTransform xform) {
        if ((xform == null) || xform.isIdentity()) {
            nSetWorldTransformToIdentity(pContext);
        } else {
            nSetWorldTransform(pContext,
                    xform.getMxx(), xform.getMxy(), xform.getMxz(), xform.getMxt(),
                    xform.getMyx(), xform.getMyy(), xform.getMyz(), xform.getMyt(),
                    xform.getMzx(), xform.getMzy(), xform.getMzz(), xform.getMzt(),
                    0.0, 0.0, 0.0, 1.0);
        }
    }

    @Override
    protected void updateClipRect(Rectangle clipRect) {

    }

    @Override
    protected void updateCompositeMode(CompositeMode mode) {

    }

    @Override
    public void blit(RTTexture srcRTT, RTTexture dstRTT, int srcX0, int srcY0, int srcX1, int srcY1, int dstX0, int dstY0, int dstX1, int dstY1) {

    }

    @Override
    protected void renderQuads(float[] coordArray, byte[] colorArray, int numVertices) {

    }

    /**
     * WIN32 COM bool FAILED(HRESULT hr) macro equivalent.
     */
    private static boolean FAILED(int hr) {
        return hr < 0;
    }

    /**
     * Does D3D11 native return value validation for debugging purposes.
     */
    static void validate(int res) {
        if (PrismSettings.verbose && FAILED(res)) {
            System.out.println("D3D11 hresult failed :" + hResultToString(res));
            new Exception("Stack trace").printStackTrace(System.out);
        }
    }

    private static String hResultToString(long hResult) {
        switch ((int) hResult) {
            /*
            case (int)D3DERR_DEVICENOTRESET:
                return "D3DERR_DEVICENOTRESET";
            case (int)D3DERR_DEVICELOST:
                return "D3DERR_DEVICELOST";
            case (int)D3DERR_OUTOFVIDEOMEMORY:
                return "D3DERR_OUTOFVIDEOMEMORY";
            case (int)D3D_OK:
                return "D3D_OK";
            */
            default:
                return "D3D11_ERROR " + Long.toHexString(hResult);
        }
    }

    private static native int nSetRenderTarget(long pContext, long pDest, boolean depthBuffer, boolean msaa);
    private static native void nSetWorldTransformToIdentity(long pContext);
    private static native void nSetWorldTransform(long pContext,
                                                  double m00, double m01, double m02, double m03,
                                                  double m10, double m11, double m12, double m13,
                                                  double m20, double m21, double m22, double m23,
                                                  double m30, double m31, double m32, double m33);
}
